### ISSUE TITLE
smplify after model prediction generate erroneous results

### ISSUE THREAD

**Issue by iammarvelous** [nkolot/SPIN#37]
I tried to run SMPLify after model prediction in the demo code after [here](https://github.com/nkolot/SPIN/blob/master/demo.py#L124)

```
kpts = np.zeros((1, 49, 3))
kpts[0, :25, :] = keypoints # keypoints load from openpose json file
pred_cam_t = torch.stack([pred_camera[:, 1],
                          pred_camera[:, 2],
                          2 * constants.FOCAL_LENGTH / (constants.IMG_RES * pred_camera[:, 0] + 1e-9)], dim=-1)


bs = 1
pred_rotmat_hom = torch.cat(
    [pred_rotmat.detach().view(-1, 3, 3).detach(),
     torch.tensor([0, 0, 1], dtype=torch.float32, device=device).view(1, 3, 1).expand(bs * 24, -1, -1)],
dim=-1)
pred_pose = rotation_matrix_to_angle_axis(pred_rotmat_hom).contiguous().view(bs, -1)
pred_pose[torch.isnan(pred_pose)] = 0.0

smplify = SMPLify(step_size=1e-2, batch_size=bs, num_iters=1000, focal_length=constants.FOCAL_LENGTH)
new_opt_vertices, new_opt_joints, \
new_opt_pose, new_opt_betas, \
new_opt_cam_t, new_opt_joint_loss = smplify(
    pred_pose.detach(), pred_betas.detach(),
    pred_cam_t.detach(),
    0.5 * constants.IMG_RES * torch.ones(bs, 2, device=device),
    keypoints)
new_opt_joint_loss = new_opt_joint_loss.mean(dim=-1)
pred_vertices = new_opt_vertices[0].cpu().numpy()
```
I rendered with the above pred_vertices and get the following weird results:
![im1010_shape](https://user-images.githubusercontent.com/14206232/71712410-1ad90100-2dd3-11ea-84e7-4f3d659f325c.png)
![im1010_shape_side](https://user-images.githubusercontent.com/14206232/71712418-2298a580-2dd3-11ea-8980-3a2397f0b83e.png)
I think the only difference between this and the training code is --- the ground truth `keypoints[25:]` is unknown here. But according to [this](https://github.com/nkolot/SPIN/issues/30#issuecomment-570276278), the unknown ground truth should not be the reason, right? Did I miss anything here?

**Comment by nkolot** [comment]
There are 2 issues with your code above:
1. You need to make sure that the keypoints are in the expected format.
2. You need to use the camera translation that SMPLify gives you for the visualization.

You can make it work if you run `python3 demo.py --img=examples/im1010.jpg --openpose=examples/im1010_openpose.json --checkpoint=data/model_checkpoint.pt`

and do the following transformation to the keypoints returned:
```
keypoints[:, :2] -= np.reshape(center, (1,2))
keypoints[:, :2] /= (scale * 200)
keypoints[:, :2] += 0.5
keypoints[:, :2] *= constants.IMG_RES
```

Essentially you need to subtract the bounding box center and divide by the size of the box (this way all the keypoints will be in the [-0.5, 0.5] range) and then add 0.5 and multiply by the standard box size (224 in our case). You *should* use `new_opt_cam_t` as your new camera translation. If you do all these changes your output should look like this:

![im1010_shape](https://user-images.githubusercontent.com/31786013/71767375-fcc2ec00-2f0b-11ea-9305-ba721da6da13.png)

![im1010_shape_side](https://user-images.githubusercontent.com/31786013/71767378-02203680-2f0c-11ea-8947-9ce7948cfac0.png)

For your reference, I run SMPLify for 30 iterations.

**Comment by iammarvelous** [comment]
Thanks. I am able to get it to work now. Just two more questions about the [implementation](https://github.com/nkolot/SPIN/blob/master/train/trainer.py#L164-L166) of `pred_camera`. It seems that the model outputs [z, x, y] instead of [x, y, z]. Is there a particular reason for the model to regress [z, x, y] instead of [x, y, z]? It seems unnecessary to me since the rest of the code only uses [x, y, z]. Also, it's not obvious to me why you want to have this [transform]( https://github.com/nkolot/SPIN/blob/master/train/trainer.py#L166). It looks like the model is regressing something like `1/z`.

**Comment by nkolot** [comment]
For the network part, we regress `[s,tx,ty]` as in HMR. These originally correspond to an orthographic scaling factor and a 2D translation `[tx,ty]`. Here we convert this orthographic representation to an actual 3D translation assuming that we have a known focal length. So as you said, s is inversely proportional to the depth `z`.

**Comment by longbowzhang** [comment]
Hi Nikos @nkolot , If i understand correctly, [tx, ty] indicates the translation in image space, then how could you manange to convert [s, tx, ty] to a translation in 3D space?

**Comment by penincillin** [comment]
@longbowzhang 
For (2), I think it's just due to the inconsistent definition of axis. It's not a big issue .
For (1), I think the authors just follow the practice of HMR, using s*tx and s*ty should also be feasible.
BTW, please be noted that I am not the author and just share my thoughts here, hope it can help you :)
